cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -2660,60 +2616,112 @@ int getPodIndexFromPvcName(String pvcName)  {
+                    if (userOperator.getLogging() instanceof ExternalLogging) {
+                        return getEntityUserOperatorDescription(configAnnotation, annotations, entityOperator);
+                    }
+                }
+            }
+            return Future.succeededFuture(this);
+        }
+
+        @SuppressWarnings("deprecation")
+        private Future<ReconciliationState> getEntityTopicOperatorDescription(final String[] configAnnotation, Map<String, String> annotations, EntityOperator entityOperator) {
+            EntityTopicOperator topicOperator = entityOperator.getTopicOperator();
+            return configMapOperations.getAsync(kafkaAssembly.getMetadata().getNamespace(), ((ExternalLogging) topicOperator.getLogging()).getName())
+                    .compose(cm -> {
+                        ConfigMap topicOperatorLogAndMetricsConfigMap = topicOperator.generateMetricsAndLogConfigMap(cm);
+                        this.topicOperatorMetricsAndLogsConfigMap = topicOperator.generateMetricsAndLogConfigMap(topicOperatorLogAndMetricsConfigMap);
+
+                        configAnnotation[0] += topicOperatorLogAndMetricsConfigMap.getData().get("log4j2.properties");
+
+                        annotations.put(io.strimzi.operator.cluster.model.TopicOperator.ANNO_STRIMZI_IO_LOGGING, configAnnotation[0]);
+
+                        this.entityOperator = entityOperator;
+                        this.eoDeployment = entityOperator.generateDeployment(pfa.isOpenshift(), annotations, imagePullPolicy, imagePullSecrets);
+
+                        return Future.succeededFuture(null);
+                    });
+        }
+
+        @SuppressWarnings("deprecation")
+        private Future<ReconciliationState> getEntityUserOperatorDescription(final String[] configAnnotation, Map<String, String> annotations, EntityOperator entityOperator) {
+            EntityUserOperator userOperator = entityOperator.getUserOperator();
+            return configMapOperations.getAsync(kafkaAssembly.getMetadata().getNamespace(), ((ExternalLogging) userOperator.getLogging()).getName())
+                    .compose(cm -> {
+                        ConfigMap userOperatorLogAndMetricsConfigMap = userOperator.generateMetricsAndLogConfigMap(cm);
+                        this.userOperatorMetricsAndLogsConfigMap = userOperator.generateMetricsAndLogConfigMap(userOperatorLogAndMetricsConfigMap);
+
+                        configAnnotation[0] += userOperatorLogAndMetricsConfigMap.getData().get("log4j2.properties");
+
+                        annotations.put(io.strimzi.operator.cluster.model.TopicOperator.ANNO_STRIMZI_IO_LOGGING, configAnnotation[0]);
+
+                        this.entityOperator = entityOperator;
+                        this.eoDeployment = entityOperator.generateDeployment(pfa.isOpenshift(), annotations, imagePullPolicy, imagePullSecrets);
+
+                        return Future.succeededFuture(this);
+                    });
```

Comment:
I didn't actually run any tests with it. But I wonder if you can rewrite this whole thing with something like this?

```java
private final Future<ReconciliationState> getEntityOperatorDescription2() {
            this.entityOperator = EntityOperator.fromCrd(kafkaAssembly, versions);

            if (entityOperator != null) {
                EntityTopicOperator topicOperator = entityOperator.getTopicOperator();
                EntityUserOperator userOperator = entityOperator.getUserOperator();

                Future<ConfigMap> futToConfigMap;

                if (topicOperator != null && topicOperator.getLogging() instanceof ExternalLogging)  {
                    futToConfigMap = configMapOperations.getAsync(kafkaAssembly.getMetadata().getNamespace(), ((ExternalLogging) topicOperator.getLogging()).getName());
                } else {
                    futToConfigMap = Future.succeededFuture(null);
                }

                Future<ConfigMap> futUoConfigMap;

                if (userOperator != null && userOperator.getLogging() instanceof ExternalLogging)  {
                    futUoConfigMap = configMapOperations.getAsync(kafkaAssembly.getMetadata().getNamespace(), ((ExternalLogging) userOperator.getLogging()).getName());
                } else {
                    futUoConfigMap = Future.succeededFuture(null);
                }

                return CompositeFuture.join(futToConfigMap, futUoConfigMap)
                        .compose(res -> {
                            ConfigMap toCm = res.resultAt(0);
                            ConfigMap uoCm = res.resultAt(1);
                            String configAnnotation = "";

                            if (topicOperator != null && toCm != null)  {
                                this.topicOperatorMetricsAndLogsConfigMap = topicOperator.generateMetricsAndLogConfigMap(toCm);
                                configAnnotation += this.topicOperatorMetricsAndLogsConfigMap.getData().get("log4j2.properties");
                            }

                            if (userOperator != null && uoCm != null)   {
                                this.userOperatorMetricsAndLogsConfigMap = userOperator.generateMetricsAndLogConfigMap(uoCm);
                                configAnnotation += this.userOperatorMetricsAndLogsConfigMap.getData().get("log4j2.properties");
                            }

                            Map<String, String> annotations = new HashMap<>();
                            annotations.put(io.strimzi.operator.cluster.model.TopicOperator.ANNO_STRIMZI_IO_LOGGING, configAnnotation);

                            this.eoDeployment = entityOperator.generateDeployment(pfa.isOpenshift(), annotations, imagePullPolicy, imagePullSecrets);

                            return Future.succeededFuture(this);
                        });
            } else {
                return Future.succeededFuture(this);
            }
        }
```